mlflow/pipelines/steps/train.py

```diff
@@ -124,135 +125,149 @@ def _run(self, output_directory):
-                )
-                if os.path.exists(output_model_path) and os.path.isdir(output_model_path):
-                    shutil.rmtree(output_model_path)
-                mlflow.sklearn.save_model(model, output_model_path)
-
-                with open(os.path.join(output_directory, "run_id"), "w") as f:
-                    f.write(run.info.run_id)
-                log_code_snapshot(
-                    self.pipeline_root, run.info.run_id, pipeline_config=self.pipeline_config
+                # wrap training in objective fn
+                def objective(hyperparameter_args):
+                    with mlflow.start_run(tags=tags, nested=True):
+                        estimator = estimator_fn(
+                            dict(estimator_hardcoded_params, **hyperparameter_args)
+                        )
+
+                        sample_fraction = self.step_config["sample_fraction"]
+                        X_train_sampled = X_train.sample(frac=sample_fraction, random_state=42)
+                        y_train_sampled = y_train.sample(frac=sample_fraction, random_state=42)
+
+                        estimator.fit(X_train_sampled, y_train_sampled)
+
+                        logged_estimator = self._log_estimator_to_mlflow(estimator, X_train_sampled)
+
+                        eval_result = mlflow.evaluate(
+                            model=logged_estimator.model_uri,
+                            data=validation_df,
+                            targets=self.target_col,
+                            model_type="regressor",
+                            evaluators="default",
+                            dataset_name="validation",
+                            custom_metrics=_load_custom_metric_functions(
+                                self.pipeline_root,
+                                self.evaluation_metrics.values(),
+                            ),
+                            evaluator_config={
+                                "log_model_explainability": False,
+                            },
+                        )
+
+                        # return +/- metric
+                        return eval_result.metrics[self.primary_metric]
+
+                search_space = self._construct_search_space_from_yaml(tuning_params["parameters"])
+                algo_type, algo_name = tuning_params["algorithm"].rsplit(".", 1)
+                tuning_algo = getattr(importlib.import_module(algo_type, "hyperopt"), algo_name)
+                max_trials = tuning_params["max_trials"]
+                hp_trials = Trials()
+
+                best_hp_params = fmin(
+                    objective,
+                    search_space,
+                    algo=tuning_algo,
+                    max_evals=max_trials,
+                    trials=hp_trials,
                 )
+                best_hp_estimator_loss = hp_trials.best_trial["result"]["loss"]
+                hardcoded_estimator_loss = objective({})
```

Comment:
This won't work if the user-defined model returned by `estimator_fn` initializes model parameters using the args passed to `estimator_fn`. We should instead call `objective()` with the hardcoded parameter dict, rather than an empty dict.